### ISSUE TITLE
`$.extend`: add support for the `deep` parameter

### ISSUE THREAD

**Issue by sntran** [fabiospampinato/cash#196]
`jQuery.extend` API supports passing a boolean argument as first param and it will recursively copy properties of the source object.

Would this library consider support that as well?

**Comment by fabiospampinato** [comment]
If someone is willing to push a PR implementing it, that doesn't add many bytes to the minified and gzipped bundle then yes, I'll probably accept it.

What is your use case for this exactly?

**Comment by fabiospampinato** [comment]
Aligning cash more with jQuery is important (https://github.com/kenwheeler/cash/issues/193), but that is more about ensuring the things we implement work the same way as they do in jQuery, we are not trying to reimplement the whole of jQuery itself.

**Comment by DataTables** [comment]
I'm the author of the DataTables plug-in for jQuery and have just been having a look at how feasible it would be to have DataTables running on cash. This was the first issue that I ran into - I'll try to make some time to take a look at implementing it.

**Comment by fabiospampinato** [comment]
@DataTables That's awesome! I'm actually still having to use jQuery in a project of mine on the pages where a DataTables table is used, it'd be great if I could ditch jQuery entirely.

>  I'll try to make some time to take a look at implementing it.

For the time being I think you should be able to just copy [Zepto's implementation](https://github.com/madrobby/zepto/blob/763b3d6dc3b4350759ed30aa196cd2b6e39efcfb/src/zepto.js#L235-L257), with very minor changes. Later if you'd like to submit a PR for this I'll be happy to consider it. 

A few recommendations: 

- cash v3 isn't released yet (I'd like to migrate to it myself in a production project of mine first), but that's the version you should use, you can get it from [here](https://cdn.jsdelivr.net/npm/cash-dom@3.0.0-beta.3/dist/cash.min.js) or via `npm i cash-dom@3.0.0-beta.3`.
- Make sure to read our [migration guide](https://github.com/kenwheeler/cash/blob/master/docs/migration_guide.md), there are few incompatibilities between cash and jquery.
- The single biggest feature from cash might be ajax support, I'd recommend taking a look at [unfetch](https://github.com/developit/unfetch) as a possible replacement. [Here](https://github.com/svelto/svelto/blob/master/src%2Flib%2Ffetch%2Ffetch.js) there's also an alternative fetch method I made a while ago, with some minor adaptations I think it could be used as a drop-in replacement for jQuery's $.ajax, for most scenario.

**Comment by DataTables** [comment]
Just trying to compile cash with a few changes I've made to support deep extend but I'm getting the following (note even on a clean checkout this is the case without my changes):

```
$ npm run build:prod

> cash-dom@3.0.0-beta.3 build:prod /home/vagrant/cash
> pacco build-javascript

[13:21:32] Starting 'build-javascript'...
[Error] Plugin `Dependencies` encountered an error
[Message] Circular dependencies found involving:
  /home/vagrant/cash/src/index.ts
  /home/vagrant/cash/src/manipulation/index.ts
  /home/vagrant/cash/src/manipulation/replace_all.ts
  /home/vagrant/cash/src/manipulation/replace_with.ts
[13:21:33] Finished 'build-javascript' after 132ms
```

Any ideas?

**Comment by fabiospampinato** [comment]
@DataTables I think I've broken the bundler [here](https://github.com/fabiospampinato/pacco/commit/f03b9070d5f627b86c2eb275eff618acd604da61), working on a fix right now.

In the mean time you may use an older version of the bundler `npm i pacco@2.0.6`.

**Comment by fabiospampinato** [comment]
@DataTables just fixed the problem, you should now be able to compile cash successfully, let me know if you encounter any other issues 👍

**Comment by DataTables** [comment]
Awesome - thanks.

I've added the following in [this commit](https://github.com/DataTables/cash/commit/5ebabbd08fbffc3685be954939721e60fbbc5315):

* cash.inArray
* cash.merge
* cash.map (array only currently, like cash.each)
* cash.isPlainObject
* cash.trim
* Deep copying for cash.extend

Those were all needed for DataTables to run. I've also [committed a couple of changes](https://github.com/DataTables/DataTablesSrc/commit/f13884ff45226784643c87a2c02ef8b1926b9411) to my v2 branch of DataTables (I'm fairly sure most of the same changes could be applied to 1.10) to address one or two points that wouldn't be suitable for cash (e.g. using `:eq()` in a selector).

I think it takes the size up from 14.4KB to around 15.3KB - so roughly a K more. Happy to send a pull request with that diff if you think its suitable?

**Comment by fabiospampinato** [comment]
@DataTables We are trying to keep cash as small as possible, ideally the minified and gzipped code should be under 5kb total. 

As a rule of thumb those sorts of utilities are not provided by cash, unless we also need them internally and jQuery exposes them too.

So I would avoid including the following methods:

- cash.inArray
- cash.merge
- cash.isPlainObject
- cash.trim

> cash.map (array only currently, like cash.each)

This one is useless to us, since we support only relatively modern browsers, so the `Array.prototype.map` method is available to us.

> Deep copying for cash.extend

Feel free to submit a PR about this. Try to keep the code as short as possible.

> I think it takes the size up from 14.4KB to around 15.3KB - so roughly a K more.

That unminified and ungzipped size doesn't really matter for us, we care about the minified and gzipped size.

Since we don't ship some of the utilities DataTables needs how do you plan to add compatibility with cash? Will you add a `cash-compat` library that includes those missing utilities maybe? 

IMHO ideally cash should be a drop-in replacement for jQuery, maybe you could consider avoiding using those missing methods entirely.

**Comment by DataTables** [comment]
The gzip version comes out at 5331 bytes.

> This one is useless to us, since we support only relatively modern browsers, so the Array.prototype.map method is available to us.

Its more as a proxy so libraries like DataTables don't need to change their code (and jQuery can be dropped in for support for older browsers).

> Since we don't ship some of the utilities DataTables needs how do you plan to add compatibility with cash?

I'm not sure I will add compatibility with Cash if it doesn't provide a drop in replacement for jQuery. I don't mind changing some things in DataTables (such as one or two of the selectors I used not being compatible with qsa) but if there are utility methods missing such as `isPlainObject` I'd just need to implement them in DataTables, so there would be no size benefit to the end user (indeed, it would result in a larger size for those using jQuery).

**Comment by fabiospampinato** [comment]
> I'm not sure I will add compatibility with Cash if it doesn't provide a drop in replacement for jQuery.

> Its more as a proxy so libraries like DataTables don't need to change their code (and jQuery can be dropped in for support for older browsers).

I understand that, but the only way to make cash a drop-in replacement for jQuery on our end is to re-implement jQuery, which we are not interested in doing.

Library authors on the other end can use only the commonly available APIs to make their libraries more compatible. Also keep in mind that cash's goal is to be "absurdly small", implementing things like `$.map`, which basically nobody needs except people supporting very old browsers (and cash itself is "only" IE10+ compatible), goes against that.

> so there would be no size benefit to the end user (indeed, it would result in a larger size for those using jQuery).

It really depends, let's say the hypothetical `cash-compact` shim that implements a few missing methods weighs .5kb minified and gzipped, then:

- DataTables + cash + cash-compact = 28kb + 5kb + .5kb = 33.5kb
- DataTables + jQuery 3 = 28kb + 30.3kb = 58.3kb
- DataTables + jQuery 3 + cash-compact = 28kb + 30.3kb + .5kb = 58.8kb

If you implement those methods in a separate bundle there's no extra cost for jQuery users, and a potential 24.8kb of savings for cash users.

If you implement those methods inside DataTables then users can more easily switch to cash and save bytes, at the cost of 1% in the library size increase, which IMHO it's probably an overly pessimistic estimate, considering that I'm happy to add deep extend support to cash. 

If all the extra methods end-up weighting .1kb once minified and gzipped, which seems a more reasonable estimate to me, then we are talking about a potential 0.2% increase in the bundle size for jQuery users for a potential 43% savings when switching to cash, and we are not even accounting for the extra time needed for parsing, compiling and executing all that extra JS that comes with jQuery.

IMHO if I add deep extend support here, and you add the other methods on your end, the value proposition is pretty good, what do you think?

**Comment by DataTables** [comment]
The deep extend needs `isPlainObject` to be added unfortunately, so I've exposed that along with the other similarly exposed jQuery methods.

I guess my issue as a library author is that I don't really want to write and support the missing methods of jQuery. I could, and it would be a way out of this, but it would also mean that any other library authors that wanted to add support for cash would also need to fill any gaps in the api. To my mind, utility functions such as this need to be in a utility library such as cash or jQuery as a core point.

I absolutely take the point about not reimplementing jQuery - that has always been my reply when folks ask me if there is going to be a DataTables version without the requirement of jQuery. The problem is I used a lot of the jQuery utility methods to avoid browser issues in the old days (and I still intend to support older browsers) - so only having a select subset makes it difficult for me and I think would limit other jQuery plug-ins from using cash.

**Comment by fabiospampinato** [comment]
> The deep extend needs isPlainObject to be added unfortunately

We could add `$.isPlainObject` too in that case.

>  I could, and it would be a way out of this, but it would also mean that any other library authors that wanted to add support for cash would also need to fill any gaps in the api. 

No way around that unfortunately, only jQuery implements everything that jQuery provides.

> To my mind, utility functions such as this need to be in a utility library such as cash or jQuery as a core point.

IMHO they made a mistake exposing some of those methods. Maybe it made sense in the old days, when browsers didn't implement some core methods like `map` or `trim` and when things like WebPack weren't a thing, but now we are left with these kinds of problems.

>  so only having a select subset makes it difficult for me and I think would limit other jQuery plug-ins from using cash.

Hopefully not too difficult, if with a few changes and less than 50 lines of code you can get something as big as DataTables to use cash. Unfortunately supporting some of those [utility methods](https://api.jquery.com/category/utilities/) just don't make sense for cash, we aren't trying to be maximally compatible with all existing plugins.

**Comment by DataTables** [comment]
I think its remarkable what you've been able to do with cash. As you say, getting something like DataTables to run with it, with just a few hours of effort on my part is a huge testament to the capability of cash and its jQuery interoperability.

The deep clone is available in the fork I linked to above with the diff if you want to pull it in that way or I can clean my fork and add just that in.

For the moment I don't think I'll be adding the extras needed to DataTables to make it compatible but I'll keep cash in mind for future development.

**Comment by Daijobou** [comment]
I request this feature too.

I test a little cash as replacement for jquery. jquery plugins use $.extend() with boolean too. Cash break here with a error:

cash.js:166 Uncaught TypeError: Cannot create property 'myvalue' on boolean 'true'

jquery code:
this.var = $.extend(true, {}, this.test[""]);

Sources:
https://github.com/kenwheeler/cash/blob/master/dist/cash.js#L153
https://github.com/jquery/jquery/blob/master/src/core.js#L118

**Comment by fabiospampinato** [comment]
Implemented in #347
